rustls/src/conn.rs

```diff
@@ -1213,32 +1243,12 @@ impl<Data> ConnectionCore<Data> {
         Ok((secrets, external))
     }
 
-    pub(crate) fn export_keying_material<T: AsMut<[u8]>>(
-        &self,
-        mut output: T,
-        label: &[u8],
-        context: Option<&[u8]>,
-    ) -> Result<T, Error> {
-        if output.as_mut().is_empty() {
-            return Err(Error::General(
-                "export_keying_material with zero-length output".into(),
-            ));
-        }
-
-        // promote fused error
-        if let Err(err) = &self.state {
-            return Err(err.clone());
+    pub(crate) fn exporter(&mut self) -> Result<KeyingMaterialExporter, Error> {
+        match self.common_state.exporter.take() {
+            Some(inner) => Ok(KeyingMaterialExporter { inner }),
+            None if self.common_state.is_handshaking() => Err(Error::HandshakeNotComplete),
+            None => Err(Error::General("exporter already used".into())),
```

Comment:
This feels like it _might_ be something worth its own error type. Did you consider that and decide it was overkill?

Reply:
Sounds nice.

Reply:
Will do this in a follow-up